Mathlib/Algebra/Order/GroupWithZero/WithZero.lean

```diff
@@ -0,0 +1,57 @@
+/-
+Copyright (c) 2024 Kevin Buzzard. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Buzzard
+-/
+import Mathlib.Algebra.Order.Ring.Lemmas
+import Mathlib.Algebra.Order.GroupWithZero.Canonical
+/-!
+
+# Covariant instances on `WithZero`
+
+Adding a zero to a type with a preorder and multiplication which satisfies some
+axiom, gives us a new type which satisfies some variant of the axiom.
+
+## Example
+
+If `α` satisfies `b₁ < b₂ → a * b₁ < a * b₂` for all `a`,
+then `WithZero α` satisfies `b₁ < b₂ → a * b₁ < a * b₂` for all `a > 0`,
+which is `PosMulStrictMono (WithZero α)`.
+
+## Application
+
+The type `ℤₘ₀ := WithZero (Multiplicative ℤ)` is used a lot in mathlib's valuation
+theory. These instances enable lemmas such as `mul_pos` to fire on `ℤₘ₀`.
+
+## TODO
+
+Add MulPosStrictMono and PosMulMono instances.
+-/
```

Comment:
```suggestion
-/

assert_not_exists Ring
```

Reply:
I only eat rings in my soup

(`Semiring` and `Ring` are defined in the same file)